File: Python/game/strategies/Minimax.py
# -*- coding: utf-8 -*-
"""
Created on Sat Oct  1 17:48:35 2022

@author: Swen
"""



#from Board.optim3 import Board
#
# import board
# board = board.initialize_board()

#from common.names import BLACK
# player = BLACK
# depth = 0
# maximizer = player
# max_depth = 1

from board import get_score, make_move, legal_moves, opponent

import numpy as np


def Minimax(board, player, maximizer = None, depth = 0, max_depth = 5):
    if maximizer is None: maximizer = player

    best_move = None
    if depth == max_depth:
        max_score = get_score(board) #Static evaluate current situation
    else:
        moves = legal_moves(board, player) #First find legal moves
        if moves is None:
            max_score = get_score(board)  #Static evaluate current situation
        else:
            max_score = -np.Inf #Initialize default to negative inf
        #print('Legal:', moves)

        for move in moves:
            #Play the move
            copy = board.copy()
            copy = make_move(copy, move, player)
            score, _ = Minimax(copy, opponent(player), maximizer, depth + 1, max_depth)
            #print('move:', move)
            #print('score:', score)
            if score > max_score:
                max_score = score
                best_move = move
            #print('best move:', best_move)

        #Check minimizer or maximizer:
    return -max_score, best_move

# from board import board_to_string
# print(board_to_string(board))
# player = BLACK
# depth = 0
# maximizer = player
# max_depth = 8
# %time Minimax(board, player, maximizer, depth, max_depth)
# print(board_to_string(board))

File: Cpp/strategies/minimax.cc
#include "strategies.ih"

#include<vector>
#include<string.h>
#include<iostream>

#include<unistd.h> 

Move minimax(char player, char* board, char maximizer, 
             int depth, int max_depth)
{
    Move best_move;                         // return value

    if (depth == max_depth)
    {
        best_move = {0, get_score(board)};  // static evaluate score
    }
    else
    {
        std::vector<int> moves = legal_moves(player, board);

        /*
        std::cout << "depth:" << depth << '\n';
        std::cout << '\n' << "Legal: ";
        for (int elem : moves)
        {
            std::cout << elem << ' ';
        }
        std::cout << '\n';
        sleep(1);
        */
        
        if (moves.size() != 0)
        {
            char opp = opponent(player);    // change to other player
            int best_score = -10000;         // initialize current best score

            ++depth;                        // increase the depth
            for (int move : moves)
            {
                //std::cout << '\n' <<"Testing: " << move << '\n';
                //sleep(1);
                char copy[100];                // reserve array for the copy
                strcpy(copy, board);           // copy the board to keep track

                make_move(move, player, copy); 
                //sleep(1);

                int score = minimax(opp, copy, maximizer, 
                                 depth, max_depth).value;

                if (score > best_score)
                {
                    best_score = score;         // update the best score
                    best_move = {move, score};
                }
            }

        }
        else
        {
            best_move = {0, get_score(board)};  // static evaluate score
        }
    }

    best_move.value = -best_move.value; //Return the negative value 

    /*
    if (depth < 5)
    {
        //sleep(1);
        std::cout << "depth:" << depth << '\n'; 
        std::cout << "best move: " << best_move.move << '\n';
        std::cout << "best value: " << best_move.value << '\n';
    }
    */    

    return best_move;
}